sql: cache hashed password and auth info during authentication

rafiss opened the pull request:
See individual commits
fixes #58869

---

This improves authentication performance in multiregion clusters and
improves availability when the system range leaseholder goes down.

I ran the test from https://github.com/cockroachdb/cockroach/pull/66768
and observed that there was no period of authentication unavailability.

Previously, each authentication attempt would require one read from each
of system.users and system.role_options. In multiregion clusters, this
adds quite a bit of excess latency. Also, if those tables' leaseholders
were not available, then authentication to all nodes would be affected.

This cache is meant to alleviate both problems. It's kept up to date by
updating the table descriptors' versions whenever authentication related
information is changed with CREATE/ALTER/DROP ROLE.

Release note (performance improvement): The latency of authenticating a
user has been improved by adding a cache for lookups of authentication
related information.

Release note (ops change): There is now a
server.authentication_cache.enabled cluster setting. The setting
defaults to true. When enabled, this cache stores authentication-related
data and will improve the latency of authentication attempts. Keeping
the cache up to date adds additional overhead when using the CREATE,
UPDATE, and DELETE ROLE commands. To minimize the overhead, any bulk
ROLE operations should be run inside of a transaction. To make the cache
more effective, any regularly-scheduled ROLE updates should be done all
together, rather than occurring throughout the day at all times.

Release note (security update): There is now a cache for per-user
authentication related information. The data in the cache is always kept
up-to-date because it checks if any change to the underlying
authentication tables has been made since the last time the cache was
updated. The cached data includes the user's hashed password, the
NOLOGIN role option, and the VALID UNTIL role option.

rafiss committed:
sql: cache hashed passwords during authn

This improves authentication performance in multiregion clusters and
improves availability when the system range leaseholder goes down.

Previously, each authentication attempt would require one read from each
of system.users and system.role_options. In multiregion clusters, this
adds quite a bit of excess latency. Also, if those tables' leaseholders
were not available, then authentication to all nodes would be affected.

This cache is meant to alleviate both problems. It's kept up to date by
updating the table descriptors' versions whenever authentication related
information is changed with CREATE/ALTER/DROP ROLE.

Release note (performance improvement): The latency of authenticating a
user has been improved by adding a cache for lookups of authentication
related information.

Release note (security update): There is now a cache for per-user
authentication related information. The data in the cache is always kept
up-to-date because it checks if any change to the underlying
authentication tables has been made since the last time the cache was
updated. The cached data includes the user's hashed password, the
NOLOGIN role option, and the VALID UNTIL role option.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/66919)
<!-- Reviewable:end -->

rafiss committed:
sql: refactor return params of password retrieval function

Release note: None

rafiss committed:
sql: move user login timeout above cache

Release note: None

knz reviewed (commented):
This is pretty good, I like it.

There's one thing that gives me some anxiety though: the fact we're increasing the descriptor version every time there is a change on *any* user.

A large deployment may have tens of thousands of users. Then we have customers that have periodic password changes policies (e.g. renew every 30 days).

There are two issues with this:

- we'll need to make it very clear  what happens  in deployment docs. I would recommend adding three release notes, not just 1: 
  -  a "security update" (explain that caching is taking place, and be specific about the condition for invalidation);  and also 
  - a "ops update" (explain that IT policies to refresh passwords should be run as batched transactions that update many users at once, and not one user at a time, and if possible do this just once per day instead of throughout the day)

- the descriptor version type is a uint32. With many users and regular password updates, we'll run out of version values sooner than later. @ajwerner what do you think about this?

> Reviewed 1 of 1 files at r1, 9 of 9 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-:-MdH6lHK5XNdLMgwGpMr:b3mjzbn)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @rafiss, and @RichardJCai)

___
*[pkg/sql/user.go, line 96 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-MdH0sMr6n4BiQbXXsjz:-MdH0sMr6n4BiQbXXsk-:bbhhah3) ([raw file](https://github.com/cockroachdb/cockroach/blob/b71055a530557fb21bb53a4170c56eaf8be53c3c/pkg/sql/user.go#L96)):*
> ```Go
> }
> 
> type passwordInfo struct {
> ```

If this struct is meant to contain the booleans `userExists` and `canLogin`, then it's not good to call it `passwordInfo` - we have plenty of users who can log in without a password (e.g. with TLS client certs).

Maybe call this `userInfo` instead.
___
*[pkg/sql/user.go, line 104 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-MdH4yoODnjSl8MdPACH:-MdH4yoODnjSl8MdPACI:bqtfvh8) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e3d93a3f8fa8db82f7fe36c18cf3082dd7f0cef/pkg/sql/user.go#L104)):*
> ```Go
> 
> // PasswordCache is a shared cache for hashed passwords and related information.
> type PasswordCache struct {
> ```

ditto my comment in the previous commit: this is likely a `UserInfoCache` not `PasswordCache`.
___
*[pkg/sql/user.go, line 162 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-MdH4gsu-_EsDWGhzjxU:-MdH4gsu-_EsDWGhzjxV:bnf9gix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e3d93a3f8fa8db82f7fe36c18cf3082dd7f0cef/pkg/sql/user.go#L162)):*
> ```Go
> 				// We release the lock here instead of using defer as we need to keep
> 				// going and re-lock if adding the looked-up entry.
> 				passwordCache.Lock()
> ```

tip: you can make the lock scope clearer by using a closure inside the loop, like this:

```go
for {
  pInfo, ok := func() (...) {
        passwordCache.Lock()
        defer passwordCache.Unlock()
        ...
  }
```

ditto below


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> the descriptor version type is a uint32. With many users and regular password updates, we'll run out of version values sooner than later. @ajwerner what do you think about this?

We can trivially fix this version problem. The lease table already uses `INT8`. You can redefine a proto int to be wider without risk of breaking compatibility if the values never overflowed the old size. I made a PR to do this (https://github.com/cockroachdb/cockroach/pull/66963). 

I'm okay with the approach. I'm very sad about the fact that we create jobs to wait for leases to drain but that's orthogonal and I can have @sajjadrizvi do something about that as part of other work. 

The biggest problem here is contention. For some users which update passwords frequently this will make uncontended writes into contended writes. That will hurt some users. Ideally we could turn this feature on and off. Doing that safely seems like a challenge because you need to wait for the nodes to stop using the cache before you stop incrementing the descriptor version. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-:-MdHqijd0SXU7qld9BkV:b-wz0655)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz, @rafiss, and @RichardJCai)

___
*[pkg/sql/user.go, line 104 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-MdH4yoODnjSl8MdPACH:-MdHuCwO-c5BKjIGy2Nv:b-r5lcap) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e3d93a3f8fa8db82f7fe36c18cf3082dd7f0cef/pkg/sql/user.go#L104)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto my comment in the previous commit: this is likely a `UserInfoCache` not `PasswordCache`.
</blockquote></details>

Can we extract this into a new package with a clear interface?


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (commented):
> The biggest problem here is contention. For some users which update passwords frequently this will make uncontended writes into contended writes. That will hurt some users. Ideally we could turn this feature on and off. Doing that safely seems like a challenge because you need to wait for the nodes to stop using the cache before you stop incrementing the descriptor version.

If we do this, does it make sense to have the setting on or off by default.

Also, could we make it so setting the cluster setting bumps the table version invalidating all the current caches the cluster version being set to off would make it so we wouldn't use the cache and we could also stop incrementing the versions?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-:-MdJBkoMEsQCxiMCMZ7j:bcbms4c)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @rafiss)



<!-- Sent from Reviewable.io -->

rafiss committed:
sql/authentication: move AuthCache a new package

Release note: None

rafiss committed:
sql/authentication: add cluster setting for auth cache

Release note (ops change): There is now a
server.authentication_cache.enabled cluster setting. The setting
defaults to true. When enabled, this cache stores authentication-related
data and will improve the latency of authentication attempts. Keeping
the cache up to date adds additional overhead when using the CREATE,
UPDATE, and DELETE ROLE commands. To minimize the overhead, any bulk
ROLE operations should be run inside of a transaction. To make the cache
more effective, any regularly-scheduled ROLE updates should be done all
together, rather than occurring throughout the day at all times.

rafiss reviewed (commented):
Added more release notes. I added some of the requests in new commits to keep this easier to review.


I added a cluster setting for this that defaults to true. How important is the requirement to wait for nodes to stop using the cache before stopping the incrementing of the table version? I tried using the cluster setting's `SetOnChange` callback, but there wasn't a clear way for it to get a handle on a `planner` so it could do @richardjcai's idea. Any tips? Another option is to modify `func (p *planner) SetClusterSetting` so that it does the table version bumping when this setting is changed (seems quite hacky),

Also, do I need to add memory accounting here? It would seem prudent, but I noticed that the RoleMemberCache does not have memory accounting.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-:-MdQNwBj5pxl8FniQiQZ:bnc5j4y)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/sql/user.go, line 96 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-MdH0sMr6n4BiQbXXsjz:-MdJ7NW87TotDHEFgsE4:bvmtndt) ([raw file](https://github.com/cockroachdb/cockroach/blob/b71055a530557fb21bb53a4170c56eaf8be53c3c/pkg/sql/user.go#L96)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

If this struct is meant to contain the booleans `userExists` and `canLogin`, then it's not good to call it `passwordInfo` - we have plenty of users who can log in without a password (e.g. with TLS client certs).

Maybe call this `userInfo` instead.
</blockquote></details>

makes sense; fixed
___
*[pkg/sql/user.go, line 104 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-MdH4yoODnjSl8MdPACH:-MdQNnHGAbfSSlqhJvGW:b-uw89qa) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e3d93a3f8fa8db82f7fe36c18cf3082dd7f0cef/pkg/sql/user.go#L104)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Can we extract this into a new package with a clear interface?
</blockquote></details>

i've extracted it to a new package.. i hope the interface is clear, but if not, please let me know your suggestions.
___
*[pkg/sql/user.go, line 162 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-MdH4gsu-_EsDWGhzjxU:-MdLKVr_6ZSWwUs8fGrM:bhpmb1g) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e3d93a3f8fa8db82f7fe36c18cf3082dd7f0cef/pkg/sql/user.go#L162)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

tip: you can make the lock scope clearer by using a closure inside the loop, like this:

```go
for {
  pInfo, ok := func() (...) {
        passwordCache.Lock()
        defer passwordCache.Unlock()
        ...
  }
```

ditto below
</blockquote></details>

nice tip!


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> I tried using the cluster setting's SetOnChange callback, but there wasn't a clear way for it to get a handle on a planner so it could do @RichardJCai's idea

SetOnChange wouldn't work anyway because it gets messy with cross-node synchronization.

> Another option is to modify func (p *planner) SetClusterSetting so that it does the table version bumping when this setting is changed (seems quite hacky),

We already have a bunch of special cases in `(*setClusterSettingNode) startExec()`. I think this one would fit right there.

> do I need to add memory accounting here? It would seem prudent, but I noticed that the RoleMemberCache does not have memory accounting.

IMHO yes it is prudent and the fact that the RoleMemberCache doesn't have memory accounting seems like a bug. Perhaps worth fixing them together.

> Reviewed 9 of 9 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-:-MdRMMUBBi2mAeJmb6IF:b-m8uwh4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RichardJCai)



<!-- Sent from Reviewable.io -->

rafiss committed:
sql: add memory accounting to role/auth caches

These caches use memory proportional to the number of users/roles, so
it's possible for them to grow large in deployments with many users.
Memory accounting helps protect against them causing an OOM error.

Release note: None

ajwerner reviewed (commented):
Okay, here's a fresh proposal:

1) If we only cache the password hashes, then we only need to bump the version on the users table when we change a password
2) We don't necessarily need to cache the password for all users. We could have the decision to cache a password be a property of the user. 

This way we don't need to worry nearly as much about the descriptor version changes in the face of refreshed passwords if most user passwords are not cached.  

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-:-MdY4cBL2dU0vZLDKmOk:byhgjz2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @RichardJCai)



<!-- Sent from Reviewable.io -->

knz commented:
we need to cache the role options in any case to check the LOGIN option

ajwerner reviewed (commented):
> we need to cache the role options in any case to check the LOGIN option

Sure, we can also increment version when changing the `LOGIN` option (or the `cache` option I'm pitching here). Also when adding a user (which is sort of already a thing because of the role membership). 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-:-MdY6PH63iE1Jq8J3st-:b-v38k3u)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @RichardJCai)



<!-- Sent from Reviewable.io -->

RichardJCai reviewed (commented):
> We don't necessarily need to cache the password for all users. We could have the decision to cache a password be a property of the user.

I was thinking about this as well, CACHE would make a lot of sense as a role option.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-:-MdY88fyAl_7fwN1I5g9:by440iy)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)



<!-- Sent from Reviewable.io -->

rafiss commented:
> we only need to bump the version on the users table when we change a password

that's already implemented in this PR: see alter_role.go.

> We don't necessarily need to cache the password for all users. We could have the decision to cache a password be a property of the user.

I have a slight opposition to this since it makes the new improvements a harder to discover. I would prefer to make this behavior opt-out, not opt-in (for the 21.2 release). Recall that the even earlier motivation for this issue was the performance of authentication in multiregion clusters (an important issue for a large customer)

Given that CC is making multiregion easier to do, I think we'll see more complaints like that, and I'd rather address them proactively instead of having to tell people about this new option after they file a support ticket.

If a customer is so concerned about the overhead of bulk updating passwords, then using the cluster setting to opt-out seems like a good solution. Basically, I'm positing that we'll have more people concerned with authentication latency than people concerned with ALTER ROLE latency or people who do these bulk password updates.

BTW, do we already know that people are doing bulk password update, or is this a hypothetical concern? Also, do we already know that people are concerned with ALTER ROLE latency specifically?

one more thing to add to all this is that it sounds like everyone in this thread agrees that there's less risk with caching the auth info for users without a password -- i.e. for those who can only do cert-based-auth. is my understanding of everyone's opinion correct?

ajwerner commented:
FWIW the bulk updating of passwords can still be cheap if done in a single transaction; the version bump will only happen once. We need to make sure that we don't rewrite the descriptor if we've already updated it in the current transaction.

knz commented:
> BTW, do we already know that people are doing bulk password update, or is this a hypothetical concern? Also, do we already know that people are concerned with ALTER ROLE latency specifically?

The concern is hypothetical. Customers with a large number of user accounts should not be using SQL to administrate their accounts instead and lean on an external directory instead.

knz reviewed (commented):
> Reviewed 25 of 25 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66919)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[pkg/sql/authorization.go, line 382 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-MdbLgYCACXssm1knN8g:-MdbLgYCACXssm1knN8h:b-6rihq6) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d3c41dbbc45dca452cb2ceccaca1f8e7ef3486f/pkg/sql/authorization.go#L382)):*
> ```Go
> 			sizeOfEntry += sizeOfBool
> 		}
> 		err = roleMembersCache.boundAccount.Grow(ctx, sizeOfEntry)
> ```

This call is not in the correct place: the Grow call should be checked *before* the cache is populated. If the grow call fails, that means the memory budget is denying the allocation, and the cache should be prevented from gaining one more entry.
___
*[pkg/sql/authentication/cache.go, line 177 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-MdbLx0iBHOzzrW4gEjH:-MdbLx0iBHOzzrW4gEjI:b-kj4jel) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d3c41dbbc45dca452cb2ceccaca1f8e7ef3486f/pkg/sql/authentication/cache.go#L177)):*
> ```Go
> 					sizeOfEntry := len(normalizedUsername.Normalized()) +
> 						sizeOfAuthInfo + len(aInfo.HashedPassword) + sizeOfTimestamp
> 					err = a.boundAccount.Grow(ctx, int64(sizeOfEntry))
> ```

see my comment above - we check the memory account before we effectively do the allocaiton.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
Also I haven't seen where in the code the memory account is freed when the cache is emptied?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-:-MdbM5hbDMHiu6yfmApn:bj6hz0w)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)



<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> Also I haven't seen where in the code the memory account is freed when the cache is emptied?

there are calls to `boundAccount.Empty(ctx)` whenever the map is emptied. e.g. lines 136 and 142 of authentication/cache.go

but i haven't worked with memory accounting before. is that the wrong way of doing it?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-:-Mdb_RKKANg4bYsUU_k6:bkmogck)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/sql/authorization.go, line 382 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-MdbLgYCACXssm1knN8g:-Mdb_DGd6QrJKf-dOIX_:b-8gxzbt) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d3c41dbbc45dca452cb2ceccaca1f8e7ef3486f/pkg/sql/authorization.go#L382)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This call is not in the correct place: the Grow call should be checked *before* the cache is populated. If the grow call fails, that means the memory budget is denying the allocation, and the cache should be prevented from gaining one more entry.
</blockquote></details>

thanks! fixed
___
*[pkg/sql/authentication/cache.go, line 177 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-MdbLx0iBHOzzrW4gEjH:-Mdb_Eeh-OZ3v9Oaf0QS:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d3c41dbbc45dca452cb2ceccaca1f8e7ef3486f/pkg/sql/authentication/cache.go#L177)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

see my comment above - we check the memory account before we effectively do the allocaiton.
</blockquote></details>

fixed


<!-- Sent from Reviewable.io -->

knz reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with two style/robustness nits

> Reviewed 16 of 16 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-:-MdqjjvO9mYgfKRQe2sX:b-evwyx7)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rafiss)

___
*[pkg/sql/authorization.go, line 342 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-MdqjOJxCFn1pAsocEuv:-MdqjOJxCFn1pAsocEuw:b-ule13v) ([raw file](https://github.com/cockroachdb/cockroach/blob/e291f10b55f32a2f802c01d05447d81ae898ec67/pkg/sql/authorization.go#L342)):*
> ```Go
> 		// We release the lock here instead of using defer as we need to keep
> 		// going and re-lock if adding the looked-up entry.
> 		roleMembersCache.Lock()
> ```

here and below, you can enclose this into a function closure and do the `defer ...Unlock()` trick that you've already done elsewhere. That will
1) clarify what is going on
2) make this code more robust in case someone else later comes along and moves code around, or adds a new error case
___
*[pkg/sql/authentication/cache.go, line 158 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-MdqjeFcDMVa5ATxR1xW:-MdqjeFcDMVa5ATxR1xX:bmefyj8) ([raw file](https://github.com/cockroachdb/cockroach/blob/e291f10b55f32a2f802c01d05447d81ae898ec67/pkg/sql/authentication/cache.go#L158)):*
> ```Go
> 				}
> 
> 				updatedCache := func() bool {
> ```

`updatedCache, err := `
___
*[pkg/sql/authentication/cache.go, line 178 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-MdqjffS1lvesap9KzyU:-MdqjffT0QbDAFHKIEhB:b4e3khe) ([raw file](https://github.com/cockroachdb/cockroach/blob/e291f10b55f32a2f802c01d05447d81ae898ec67/pkg/sql/authentication/cache.go#L178)):*
> ```Go
> 					err = a.boundAccount.Grow(ctx, int64(sizeOfEntry))
> 					a.cache[normalizedUsername] = aInfo
> 					return true
> ```

`return true, err`


<!-- Sent from Reviewable.io -->

knz reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66919)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rafiss)

___
*[pkg/sql/authorization.go, line 382 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-MdqkAqb01241P0qB9oA:-MdqkAqb01241P0qB9oB:b-i82xa8) ([raw file](https://github.com/cockroachdb/cockroach/blob/e291f10b55f32a2f802c01d05447d81ae898ec67/pkg/sql/authorization.go#L382)):*
> ```Go
> 		if err := roleMembersCache.boundAccount.Grow(ctx, sizeOfEntry); err != nil {
> 			roleMembersCache.Unlock()
> 			return nil, err
> ```

I just realized - if the memory account says "no", that just means that the entry cannot be added to the cache, but the entry can still be used for authentication. So I don't think it should be an error. Otherwise, a server that's nearly out of memory (where the memory budgets are exhausted) will start to refuse connections.
___
*[pkg/sql/authentication/cache.go, line 176 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-MdqkQ1HCChIQ-J5pvNF:-MdqkQ1HCChIQ-J5pvNG:b-4knc2o) ([raw file](https://github.com/cockroachdb/cockroach/blob/e291f10b55f32a2f802c01d05447d81ae898ec67/pkg/sql/authentication/cache.go#L176)):*
> ```Go
> 					sizeOfEntry := len(normalizedUsername.Normalized()) +
> 						sizeOfAuthInfo + len(aInfo.HashedPassword) + sizeOfTimestamp
> 					err = a.boundAccount.Grow(ctx, int64(sizeOfEntry))
> ```

also I just realized - if the err is not nil here, you should not add the entry to the cache below :)


<!-- Sent from Reviewable.io -->

knz commented:
I apologize already for the conflicting review. The overall structure of the code is fine, we just need to get the error case right now.

RichardJCai reviewed (approved):
LGTM mod batching the schema changes for bumping the table versions.

Also didn't rigorously look over the memory accounting.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-:-Mdxnljv7uekeOi1VAPT:bqfbfg5)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rafiss)

___
*[pkg/bench/ddl_analysis/testdata/benchmark_expectations, line 2 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-MdxnalCE1gIv6lWkOK_:-MdxnalCE1gIv6lWkOKa:bad4o0c) ([raw file](https://github.com/cockroachdb/cockroach/blob/e291f10b55f32a2f802c01d05447d81ae898ec67/pkg/bench/ddl_analysis/testdata/benchmark_expectations#L2)):*
> ```
> exp,benchmark
> 18,AlterRole/alter_role_with_1_option
> ```

I'm somewhat surprised that the roundtrips increases this much, related to my other point but maybe batching the schema changes for bumping the table versions might improve this?
___
*[pkg/sql/user.go, line 286 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-Mdxn6X1AGNW5bNRIT7C:-Mdxn6X1AGNW5bNRIT7D:b-fkkjvi) ([raw file](https://github.com/cockroachdb/cockroach/blob/e291f10b55f32a2f802c01d05447d81ae898ec67/pkg/sql/user.go#L286)):*
> ```Go
> // BumpUsersTableVersion increases the table version for the
> // users table.
> func (p *planner) BumpUsersTableVersion(ctx context.Context) error {
> ```

nit: Super minor but since we always bump the Users and RoleOptions table version together for these two functions, can we turn it into one and do the schema change in one batch.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66919)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @knz, @rafiss, and @RichardJCai)

___
*[pkg/sql/authorization.go, line 342 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-MdqjOJxCFn1pAsocEuv:-Mdy-s5k7Uc7n6OLytrK:b-kk45m9) ([raw file](https://github.com/cockroachdb/cockroach/blob/e291f10b55f32a2f802c01d05447d81ae898ec67/pkg/sql/authorization.go#L342)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

here and below, you can enclose this into a function closure and do the `defer ...Unlock()` trick that you've already done elsewhere. That will
1) clarify what is going on
2) make this code more robust in case someone else later comes along and moves code around, or adds a new error case
</blockquote></details>

fully agree on these points -- i wasn't sure if it was necessary to combine into this PR, but will do it
___
*[pkg/sql/authorization.go, line 382 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-MdqkAqb01241P0qB9oA:-Mdy0Fei6qJaemWRCKII:bhhwx04) ([raw file](https://github.com/cockroachdb/cockroach/blob/e291f10b55f32a2f802c01d05447d81ae898ec67/pkg/sql/authorization.go#L382)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I just realized - if the memory account says "no", that just means that the entry cannot be added to the cache, but the entry can still be used for authentication. So I don't think it should be an error. Otherwise, a server that's nearly out of memory (where the memory budgets are exhausted) will start to refuse connections.
</blockquote></details>

hm yes very good point. so i will change this to only log the memory accounting error and allow authn to succeed
___
*[pkg/sql/user.go, line 286 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-Mdxn6X1AGNW5bNRIT7C:-Mdy0hDv4Uff6pJpg2pt:b-p1vh4i) ([raw file](https://github.com/cockroachdb/cockroach/blob/e291f10b55f32a2f802c01d05447d81ae898ec67/pkg/sql/user.go#L286)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

nit: Super minor but since we always bump the Users and RoleOptions table version together for these two functions, can we turn it into one and do the schema change in one batch.
</blockquote></details>

well, technically they aren't always bumped at the same time (e.g. if ALTER ROLE is used but the password doesn't change). but i will see what i can do
___
*[pkg/sql/authentication/cache.go, line 158 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-MdqjeFcDMVa5ATxR1xW:-Mdy16D0DnjhdZZ87bQY:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/e291f10b55f32a2f802c01d05447d81ae898ec67/pkg/sql/authentication/cache.go#L158)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`updatedCache, err := `
</blockquote></details>

fixed
___
*[pkg/sql/authentication/cache.go, line 176 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-MdqkQ1HCChIQ-J5pvNF:-Mdy1B-X14jdgN05pQ31:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/e291f10b55f32a2f802c01d05447d81ae898ec67/pkg/sql/authentication/cache.go#L176)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

also I just realized - if the err is not nil here, you should not add the entry to the cache below :)
</blockquote></details>

fixed
___
*[pkg/sql/authentication/cache.go, line 178 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-MdqjffS1lvesap9KzyU:-Mdy1Bv91oZgnqyUfPT-:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/e291f10b55f32a2f802c01d05447d81ae898ec67/pkg/sql/authentication/cache.go#L178)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`return true, err`
</blockquote></details>

fixed


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66919)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @knz, @rafiss, and @RichardJCai)

___
*[pkg/sql/authentication/cache.go, line 158 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-MdqjeFcDMVa5ATxR1xW:-Mdy3qh2AOOY_vVUsQ6w:b-6w7vtn) ([raw file](https://github.com/cockroachdb/cockroach/blob/e291f10b55f32a2f802c01d05447d81ae898ec67/pkg/sql/authentication/cache.go#L158)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

fixed
</blockquote></details>

based on the later comment (we don't want the memory accounting error to bubble up and prevent login), i won't be returning the error here.
___
*[pkg/sql/authentication/cache.go, line 178 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-MdqjffS1lvesap9KzyU:-Mdy48Ly9eJcshCYhyLW:b-6w7vtn) ([raw file](https://github.com/cockroachdb/cockroach/blob/e291f10b55f32a2f802c01d05447d81ae898ec67/pkg/sql/authentication/cache.go#L178)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

fixed
</blockquote></details>

based on the later comment (we don't want the memory accounting error to bubble up and prevent login), i won't be returning the error here.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
did you push your changes?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-:-Me-Rh8_CwS63Gjv7j78:b-i90xti)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @knz, @rafiss, and @RichardJCai)



<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
just pushed now. i was working on richard's request and resolving rebase conflicts

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-:-Me0cetzAcXbkez_Z12E:bb3u5wn)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @knz and @RichardJCai)

___
*[pkg/bench/ddl_analysis/testdata/benchmark_expectations, line 2 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-MdxnalCE1gIv6lWkOK_:-Me0cO1KDJmqaaUBwB2-:b2g9nrg) ([raw file](https://github.com/cockroachdb/cockroach/blob/e291f10b55f32a2f802c01d05447d81ae898ec67/pkg/bench/ddl_analysis/testdata/benchmark_expectations#L2)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

I'm somewhat surprised that the roundtrips increases this much, related to my other point but maybe batching the schema changes for bumping the table versions might improve this?
</blockquote></details>

i tried the batching approach, and it didn't change these numbers. (for some of them it lowered by 1, and others it increased by 1.) i'll just keep it as is, since the code is simpler than with batching


<!-- Sent from Reviewable.io -->

knz reviewed (approved):
> Reviewed 16 of 16 files at r11, 1 of 1 files at r12.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66919)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @RichardJCai)

___
*[pkg/sql/authorization.go, line 382 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-MdqkAqb01241P0qB9oA:-Me1Rual8LZnyngMVIJd:b76h66w) ([raw file](https://github.com/cockroachdb/cockroach/blob/e291f10b55f32a2f802c01d05447d81ae898ec67/pkg/sql/authorization.go#L382)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

hm yes very good point. so i will change this to only log the memory accounting error and allow authn to succeed
</blockquote></details>

Maybe explain the reasoning in a comment for the benefit of the future maintainer of this code.
___
*[pkg/sql/authentication/cache.go, line 176 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-MdqkQ1HCChIQ-J5pvNF:-Me1S08z00Q7RSufpNGZ:b4uupru) ([raw file](https://github.com/cockroachdb/cockroach/blob/e291f10b55f32a2f802c01d05447d81ae898ec67/pkg/sql/authentication/cache.go#L176)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

fixed
</blockquote></details>

also maybe explain in a comment


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66919)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @knz and @RichardJCai)

___
*[pkg/sql/authorization.go, line 382 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-MdqkAqb01241P0qB9oA:-Me1u_CYD7sxwIoE9ppi:b-4mhpgl) ([raw file](https://github.com/cockroachdb/cockroach/blob/e291f10b55f32a2f802c01d05447d81ae898ec67/pkg/sql/authorization.go#L382)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Maybe explain the reasoning in a comment for the benefit of the future maintainer of this code.
</blockquote></details>

added a comment
___
*[pkg/sql/authentication/cache.go, line 176 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66919#-MdqkQ1HCChIQ-J5pvNF:-Me1uYWd6T7gtbifz6qa:b-7tfmjp) ([raw file](https://github.com/cockroachdb/cockroach/blob/e291f10b55f32a2f802c01d05447d81ae898ec67/pkg/sql/authentication/cache.go#L176)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

also maybe explain in a comment
</blockquote></details>

added


<!-- Sent from Reviewable.io -->

rafiss commented:
thanks for the reviews everyone!

bors r=knz,RichardJCai

shermanCRL commented:
There’s interest in a backport to 20.2 here. Thoughts?
